helios-client/src/main/java/com/spotify/helios/client/HeliosClient.java

```diff
@@ -344,20 +366,34 @@ private HttpURLConnection connect(final URI uri, final String method, final byte
     throw new TimeoutException("Timed out connecting to master");
   }
 
-  private HttpURLConnection connect0(final URI uri, final String method, final byte[] entity,
-                                     final Map<String, List<String>> headers)
+  private HttpURLConnection connect0(final URI ipUri, final String method, final byte[] entity,
+                                     final Map<String, List<String>> headers,
+                                     final String hostname)
       throws IOException {
     if (log.isTraceEnabled()) {
-      log.trace("req: {} {} {} {} {} {}", method, uri,
-                headers.size(),
+      log.trace("req: {} {} {} {} {} {}", method, ipUri, headers.size(),
                 Joiner.on(',').withKeyValueSeparator("=").join(headers),
                 entity.length, Json.asPrettyStringUnchecked(entity));
     } else {
-      log.debug("req: {} {} {} {}", method, uri, headers.size(), entity.length);
+      log.debug("req: {} {} {} {}", method, ipUri, headers.size(), entity.length);
     }
 
-    final URLConnection urlConnection = uri.toURL().openConnection();
+    final URLConnection urlConnection = ipUri.toURL().openConnection();
     final HttpURLConnection connection = (HttpURLConnection) urlConnection;
+
+    // We verify the TLS certificate against the original hostname since verifying against the
+    // IP address will fail
+    if (urlConnection instanceof HttpsURLConnection) {
+      ((HttpsURLConnection) connection).setHostnameVerifier(new HostnameVerifier() {
+        @Override
+        public boolean verify(String ip, SSLSession sslSession) {
+          final String tHostname = hostname.endsWith(".") ?
+                                   hostname.substring(0, hostname.length() - 1) : hostname;
+          return new DefaultHostnameVerifier().verify(tHostname, sslSession);
+        }
+      });
+    }
+
     connection.setRequestProperty("Accept-Encoding", "gzip");
```

Comment:
Probably should add:

    System.setProperty("sun.net.http.allowRestrictedHeaders", "true");
    connection.setRequestProperty("Host", hostname);

For the case where Helios is behind nginx or some other proxy that serves multiple hostnames on the same IP.

Another way to approach this rather than creating an IP URI, etc., is to just set the request proxy to the IP:

    Proxy proxy = new Proxy(Proxy.Type.HTTP, new InetSocketAddress("10.0.0.1", 8080));
    conn = new URL(urlString).openConnection(proxy);

That might actually simplify the code quite a bit, though I'm not sure if it would work with HTTPS.